src/array.jl

```diff
@@ -836,7 +836,30 @@ unique(A::CategoricalArray{T}) where {T} = _unique(T, A.refs, A.pool)
 Drop levels which do not appear in categorical array `A` (so that they will no longer be
 returned by [`levels`](@ref DataAPI.levels)).
 """
-droplevels!(A::CategoricalArray) = levels!(A, intersect(levels(A), unique(A)))
+function droplevels!(A::CategoricalArray)
+    arefs = refs(A)
+    nlevels = length(levels(A))
+    seen = fill(false, nlevels)
+    nseen = 0
+    @inbounds for ref in arefs
+        if (ref > 0) && !seen[ref]
+            seen[ref] = true
+            nseen += 1
+            (nseen == nlevels) && return A # all levels observed, nothing to drop
+        end
+    end
+    # recode refs to keep only the seen ones (optimized version of update_refs!())
+    newlv = 0
+    levelsmap = [s ? (newlv += 1) : 0 for s in seen]
+    @inbounds for i in eachindex(arefs)
+        if arefs[i] > 0
+            arefs[i] = levelsmap[arefs[i]]
+        end
+    end
```

Comment:
0 isn't a possible value, right? Dropping the check should enable SIMD to it's probably worth it.
```suggestion
    levelsmap = [s ? (newlv += 1) : 0 for s in seen]
    @inbounds for i in eachindex(arefs)
        arefs[i] = levelsmap[arefs[i]]
    end
```

Reply:
I think 0 is possible here since it encodes missing.

Reply:
This is correct. However, performance wise, it might good to check if to storing `0` as the first element (and then dropping it) would not be faster (it would avoid branching, but would add more to indexing). If this is not the case - I think at least an `eltype` chceck can be added (so in typical case of non-missing data things are faster). I have proposed such.

Reply:
Yes, usually the pattern I used is to allocate one entry for missing values and index with `+1`.

Reply:
Ah, and actually 0 is possible even for arrays that don't allow for missing values, due to `#undef` entries. So we need the `+1`in all cases.

Reply:
I've adopted the "+1" pattern in the latest commit.